vsphere/resource_vsphere_vmfs_datastore.go

```diff
@@ -0,0 +1,324 @@
+				// resource. We need to instruct the user to remove the datastore
+				// manually.
+				return fmt.Errorf(formatCreateRollbackErrorUpdate, disk, err, remErr)
+			}
+			return fmt.Errorf("error fetching datastore extend spec for disk %s: %s", disk, err)
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+		defer cancel()
+		if _, err := extendVmfsDatastore(ctx, dss, ds, *spec); err != nil {
+			if remErr := removeDatastore(dss, ds); remErr != nil {
+				// We could not destroy the created datastore and there is now a dangling
+				// resource. We need to instruct the user to remove the datastore
+				// manually.
+				return fmt.Errorf(formatCreateRollbackErrorUpdate, disk, err, remErr)
+			}
+			return fmt.Errorf("error extending datastore with disk %s: %s", disk, err)
+		}
+	}
+
+	d.SetId(ds.Reference().Value)
+
+	// Done
+	return resourceVSphereVmfsDatastoreRead(d, meta)
+}
+
+func resourceVSphereVmfsDatastoreRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*govmomi.Client)
+	id := d.Id()
+	ds, err := datastoreFromID(client, id)
+	if err != nil {
+		return fmt.Errorf("cannot find datastore: %s", err)
+	}
+	props, err := datastoreProperties(ds)
+	if err != nil {
+		return fmt.Errorf("could not get properties for datastore: %s", err)
+	}
+	if err := flattenDatastoreSummary(d, &props.Summary); err != nil {
+		return err
+	}
+
+	// We also need to update the disk list from the summary.
+	var disks []string
+	for _, disk := range props.Info.(*types.VmfsDatastoreInfo).Vmfs.Extent {
+		disks = append(disks, disk.DiskName)
+	}
+	if err := d.Set("disks", disks); err != nil {
+		return err
+	}
+	return nil
+}
+
+func resourceVSphereVmfsDatastoreUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*govmomi.Client)
+	hsID := d.Get("host_system_id").(string)
+	dss, err := hostDatastoreSystemFromHostSystemID(client, hsID)
+	if err != nil {
+		return fmt.Errorf("error loading host datastore system: %s", err)
+	}
+
+	id := d.Id()
+	ds, err := datastoreFromID(client, id)
+	if err != nil {
+		return fmt.Errorf("cannot find datastore: %s", err)
+	}
+
+	// Veto this update if it means a disk was removed. Shrinking
+	// datastores/removing extents is not supported.
+	old, new := d.GetChange("disks")
+	for _, v1 := range old.([]interface{}) {
+		var found bool
+		for _, v2 := range new.([]interface{}) {
+			if v1.(string) == v2.(string) {
+				found = true
+			}
+		}
+		if !found {
+			return fmt.Errorf("disk %s found in state but not config (removal of disks is not supported)", v1)
+		}
+	}
+
+	// Maintain a copy of the disks that have been added just in case the update
+	// fails in the middle.
+	old2 := make([]interface{}, len(old.([]interface{})))
+	copy(old2, old.([]interface{}))
+
+	// Now we basically reverse what we did above when we were checking for
+	// removed disks, and add any new disks that have been added.
+	for _, v1 := range new.([]interface{}) {
```

Comment:
the more I read into this, the more this feels like it belongs in the SDK?

Reply:
There is no method in `Datastore` or `HostDatastoreSystem` that would encapsulate this AFAIK, and even in the vSphere Client you need to extend datastores extent by extent.

I think it would be neat if it was in the SDK - my problem with that is if we start to send feature requests or pull requests over to VMware, are now blocked by VMware's schedule for getting those changes merged upstream so that we can have clean code. I would rather implement things ourselves for now, especially while we are pushing to get a reasonable feature set in the provider, while keeping things clean enough so that if said functionality becomes available upstream, refactoring should just be as simple as a drop-in replacement.

Reply:
That makes sense - that said there's nothing stopping us wrapping this outside of the resource which would be an intermediate step. This isn't a blocker by any means, more one to think about - but I'd argue it's worth refactoring this out if possible :)